src/main/java/com/aws/greengrass/deployment/ConnectivityValidator.java

```diff
@@ -0,0 +1,254 @@
+     *
+     * @param totallyCompleteFuture Future will be updated if validation fails
+     * @throws InterruptedException if sleep is interrupted
+     */
+    public boolean httpClientCanConnect(CompletableFuture<DeploymentResult> totallyCompleteFuture)
+            throws InterruptedException {
+        logger.atDebug().log("Checking HTTP client can connect");
+
+        ImmutablePair<SdkHttpClient, GreengrassV2DataClient> pair = factory
+                .createClientFromConfig(deploymentConfiguration);
+        httpClient = pair.left;
+        greengrassV2DataClient = pair.right;
+        ListThingGroupsForCoreDeviceRequest request = ListThingGroupsForCoreDeviceRequest.builder()
+                .coreDeviceThingName(Coerce.toString(deploymentConfiguration.getThingName())).build();
+        for (int i = 0; i < maxRetries; i++) {
+            try {
+                greengrassV2DataClient.listThingGroupsForCoreDevice(request);
+                return true;
+            } catch (GreengrassV2DataException | SdkClientException e) {
+                logger.atError().cause(e).log("Http client failed to connect with the new configuration");
+                Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+            }
+        }
+        totallyCompleteFuture
+                .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE,
+                        new ComponentConfigurationValidationException(
+                                "HTTP client failed to connect to IoT Core using new configurations",
+                                DeploymentErrorCode.FAILED_TO_RECONNECT)));
+        return false;
+    }
+
+    /**
+     * Perform connectivity validation if configs have changed meaningfully.
+     *
+     * @param totallyCompleteFuture Future will be updated if validation fails
+     * @param currentDeviceConfiguration current configs the device is using
+     */
+    public boolean validate(CompletableFuture<DeploymentResult> totallyCompleteFuture,
+                            DeviceConfiguration currentDeviceConfiguration) {
+        boolean needsValidation;
+        try {
+            needsValidation = mqttClientNeedsValidation(currentDeviceConfiguration);
+            if (needsValidation && !mqttClientCanConnect(totallyCompleteFuture)) {
+                return false;
+            }
+
+            needsValidation = httpClientNeedsValidation(currentDeviceConfiguration);
+            if (needsValidation && !httpClientCanConnect(totallyCompleteFuture)) {
+                return false;
+            }
+        } catch (InterruptedException e) {
+            logger.atError().cause(e).log("Connectivity validation was unexpectedly interrupted");
+            totallyCompleteFuture
+                    .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE,
+                            new ComponentConfigurationValidationException(e,
```

Comment:
Am I correct in understanding that the logger won't log anything because the exception `e` is empty or there's no point to having `.cause(e)` because `e` is empty?